File: Data Structures & Algorithms/find-pivot-index/submission-2.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total = 0
        for num in nums:
            total += num
        
        n = len(nums)
        leftSum = 0
        for i in range(n):
            rightSum = total - leftSum - nums[i]
            if (leftSum == rightSum):
                return i
            leftSum += nums[i]
        return -1

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int l = 0, r = n-1;
        int cut = 0;

        while (l <= r) {
            int m = (r - l) / 2 + l;
            if (nums[m] > nums[r]) {
                l = m + 1;
            } else if (nums[m] < nums[r]) {
                r = m;
            } else {
                cut = m;
                break;
            }
        }
        r = n-1;
        if (target < nums[r]) {
            l = cut;
        } else if (target > nums[r]) {
            l = 0;
            r = cut - 1;
        } else {
            return r;
        }

        while (l <= r) {
            int m = (r - l) / 2 + l;
            if (nums[m] < target) {
                l = m + 1;
            } else if (nums[m] > target) {
                r = m - 1;
            } else {
                return m;
            }
        }
        return -1;
    }
};
